File: examples/kotlin/src/main/kotlin/adr/Domain.kt
// ── The inner ring: entities (seam 01) ──────────────────────────────────────
// Pure domain types and pure state transitions. No I/O, no framework, no clock.
// State is a discriminated-union value; every transition returns a NEW State.

package adr

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
enum class Actor { Human, Agent }

@Serializable
enum class Priority { Low, Normal, High, Urgent }

@Serializable
sealed interface TicketStatus {
    @Serializable @SerialName("Open") data object Open : TicketStatus
    @Serializable @SerialName("Escalating") data class Escalating(val by: Actor) : TicketStatus
    @Serializable @SerialName("Escalated") data object Escalated : TicketStatus
    @Serializable @SerialName("Resolved") data object Resolved : TicketStatus
}

@Serializable
data class Ticket(
    val id: String,
    val body: String,
    val priority: Priority,
    val status: TicketStatus,
)

@Serializable
sealed interface RunStatus {
    @Serializable @SerialName("Idle") data object Idle : RunStatus
    @Serializable @SerialName("Working") data object Working : RunStatus
    @Serializable @SerialName("Degraded") data class Degraded(val cause: String) : RunStatus
    @Serializable @SerialName("Error") data class Error(val fault: String) : RunStatus
}

// The whole world the core knows about. A value: structural equality is what
// makes replay's "fold twice, assert identical" a meaningful claim.
data class State(
    val tickets: Map<String, Ticket>,
    val focused: String?,
    val run: RunStatus,
) {
    companion object {
        fun initial(tickets: List<Ticket> = emptyList()): State =
            State(
                tickets = tickets.associateBy { it.id },
                focused = null,
                run = RunStatus.Idle,
            )
    }
}

// ── Pure transitions — copy-on-write, never mutate the input ─────────────────

fun State.withPriority(ticket: String, level: Priority): State {
    val t = tickets[ticket] ?: return withUnhandled("priority for unknown ticket $ticket")
    return copy(tickets = tickets + (ticket to t.copy(priority = level)))
}

fun State.withStatus(ticket: String, status: TicketStatus): State {
    val t = tickets[ticket] ?: return withUnhandled("status for unknown ticket $ticket")
    return copy(tickets = tickets + (ticket to t.copy(status = status)))
}

fun State.withFocus(ticket: String): State = copy(focused = ticket)

fun State.withUnhandled(note: String): State = copy(run = RunStatus.Degraded(note))

File: reference-impl/typescript/src/domain.ts
// ── The inner ring: entities (seam 01) ──────────────────────────────────────
// Pure domain types and pure state transitions. No I/O, no framework, no clock.
// State is a discriminated-union value; every transition returns a NEW State.

export type Actor = "Human" | "Agent";
export type Priority = "Low" | "Normal" | "High" | "Urgent";

export type TicketStatus =
  | { readonly kind: "Open" }
  | { readonly kind: "Escalating"; readonly by: Actor }
  | { readonly kind: "Escalated" }
  | { readonly kind: "Resolved" };

export interface Ticket {
  readonly id: string;
  readonly body: string;
  readonly priority: Priority;
  readonly status: TicketStatus;
}

export type RunStatus =
  | { readonly kind: "Idle" }
  | { readonly kind: "Working" }
  | { readonly kind: "Degraded"; readonly cause: string }
  | { readonly kind: "Error"; readonly fault: string };

export interface State {
  readonly tickets: ReadonlyMap<string, Ticket>;
  readonly focused: string | null;
  readonly run: RunStatus;
}

export function initialState(tickets: readonly Ticket[] = []): State {
  return {
    tickets: new Map(tickets.map((t) => [t.id, t])),
    focused: null,
    run: { kind: "Idle" },
  };
}

// ── Pure transitions — copy-on-write, never mutate the input ─────────────────

export function withPriority(s: State, ticket: string, level: Priority): State {
  const t = s.tickets.get(ticket);
  if (!t) return withUnhandled(s, `priority for unknown ticket ${ticket}`);
  const tickets = new Map(s.tickets);
  tickets.set(ticket, { ...t, priority: level });
  return { ...s, tickets };
}

export function withStatus(s: State, ticket: string, status: TicketStatus): State {
  const t = s.tickets.get(ticket);
  if (!t) return withUnhandled(s, `status for unknown ticket ${ticket}`);
  const tickets = new Map(s.tickets);
  tickets.set(ticket, { ...t, status });
  return { ...s, tickets };
}

export function withFocus(s: State, ticket: string): State {
  return { ...s, focused: ticket };
}

export function withUnhandled(s: State, note: string): State {
  return { ...s, run: { kind: "Degraded", cause: note } };
}
